view/frontend/layout/catalog_product_view.xml

```diff
@@ -38,16 +38,51 @@
 <page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
       xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd">
     <body>
+        <referenceContainer name="before.body.end">
+            <!-- Set page type to 'product' for TaggingProvider -->
+            <block class="Nosto\Tagging\Block\TaggingProvider" name="nosto.tagging.provider" template="Nosto_Tagging::tagging-provider.phtml">
+                <arguments>
+                    <argument name="page_type" xsi:type="string">product</argument>
+                </arguments>
+            </block>
+
+            <!-- Product tagging -->
+            <block class="Nosto\Tagging\Block\Product" name="nosto.product" after="-"/>
+
+            <!-- Category tagging (for breadcrumbs) -->
+            <block class="Nosto\Tagging\Block\Category" name="nosto.category" after="-"/>
+
+            <!-- Recommendation slots -->
+            <block class="Nosto\Tagging\Block\Element" name="nosto.page.product1" after="-"
+                   template="Nosto_Tagging::element.phtml">
+                <arguments>
+                    <argument name="nostoId" xsi:type="string">nosto-page-product1</argument>
```

Comment:
I'm not sure if the rec placements will be duplicated in the DOM.

The same elements are present twice in this XML, probably Magento will override the same element, but please test and make sure you only see that once.

Reply:
This change is not present in the other page configurations.